joblib/numpy_pickle.py

```diff
@@ -63,135 +47,428 @@ def _read_magic(file_handle):
-    """Read the z-file and return the content as a string
+def _check_filetype(filename, magic, mode='rb'):
+    """ Utility function opening and returning a fileobject from a filename
+    given it's magic number (compressed using gzip or not compressed
+
+    Parameters
+    ----------
+    filename: str
+        The full path string of the file
+    magic: bytes
+        The bytes representation of the magic number of the file
+    mode: str
+        The mode used to open the file: 'rb' (default) or 'wb'
+
+    Returns
+    -------
+        a file like object
 
-    Z-files are raw data compressed with zlib used internally by joblib
-    for persistence. Backward compatibility is not guaranteed. Do not
-    use for external purposes.
     """
-    file_handle.seek(0)
-    assert _read_magic(file_handle) == _ZFILE_PREFIX, \
-        "File does not have the right magic"
-    header_length = len(_ZFILE_PREFIX) + _MAX_LEN
-    length = file_handle.read(header_length)
-    length = length[len(_ZFILE_PREFIX):]
-    length = int(length, 16)
-
-    # With python2 and joblib version <= 0.8.4 compressed pickle header is one
-    # character wider so we need to ignore an additional space if present.
-    # Note: the first byte of the zlib data is guaranteed not to be a
-    # space according to
-    # https://tools.ietf.org/html/rfc6713#section-2.1
-    next_byte = file_handle.read(1)
-    if next_byte != b' ':
-        # The zlib compressed data has started and we need to go back
-        # one byte
-        file_handle.seek(header_length)
-
-    # We use the known length of the data to tell Zlib the size of the
-    # buffer to allocate.
-    data = zlib.decompress(file_handle.read(), 15, length)
-    assert len(data) == length, (
-        "Incorrect data length while decompressing %s."
-        "The file could be corrupted." % file_handle)
-    return data
-
-
-def write_zfile(file_handle, data, compress=1):
-    """Write the data in the given file as a Z-file.
-
-    Z-files are raw data compressed with zlib used internally by joblib
-    for persistence. Backward compatibility is not guarantied. Do not
-    use for external purposes.
+    if mode not in ('rb', 'wb'):
+        raise ValueError("Wrong file object mode, "
+                         "allowed modes are: 'rb', 'wb'")
+
+    if magic == _GZIP_PREFIX:
+        return gzip.GzipFile(filename, mode, compresslevel=3)
+
+    return open(filename, mode)
+
+
+# difference between version 1.0 and 2.0 is a 4 byte (I) header length
+# instead of 2 bytes (H) allowing storage of large structured arrays
+
+def _check_version(version):
     """
-    file_handle.write(_ZFILE_PREFIX)
-    length = hex_str(len(data))
-    # Store the length of the data
-    file_handle.write(asbytes(length.ljust(_MAX_LEN)))
-    file_handle.write(zlib.compress(asbytes(data), compress))
+    Utility that check if the version is supported. An exception is raised if
+    version is not supported.
+    Parameters
```

Comment:
Not that important at this time of the PR but it's probably worth reading https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt and fixing up the docstrings.

For example on this particular docstring: short description that fit on one-line, newline, longer description, newline, 
sections like Parameters etc ...